core/src/main/java/org/apache/iceberg/util/StructLikeMap.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.util;
+
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.Types;
+
+public class StructLikeMap<T> extends AbstractMap<StructLike, T> implements Map<StructLike, T> {
+
+  public static <T> StructLikeMap<T> create(Types.StructType type) {
+    return new StructLikeMap<>(type);
+  }
+
+  private final Types.StructType type;
+  private final Map<StructLikeWrapper, T> wrapperMap;
+  private final ThreadLocal<StructLikeWrapper> wrappers;
```

Comment:
is ThreadLocal serializable? Should StructLikeMap be?

Reply:
The `StructLikeMap` is used for locating inserted row's the file/position, pls see [here](https://github.com/apache/iceberg/pull/1818/files#diff-fc9a9fd84d24c607fd85e053b08a559f56dd2dd2a46f1341c528e7a0269f873cR117-R143).  We want to write the file/pos into positional files when encountering a delete which will delete the row that has been inserted in same txn. 

Here, the `StructLikeMap` don't have to be serializable. For flink,  we will use the [RowDataTaskWriterFactory](https://github.com/apache/iceberg/pull/1818/files#diff-a26cb8b3f0f5b0e6a9f110bb0b3fdd0ed625bfc9faf27e248f6446818cad4d69R245) to construct the IcebergStreamWriter,  we will require the factory to be serializable , the TaskWriter don't have to. So finally the `StructLikeMap`  also don't have to be serializable.